sequence_processing_pipeline/Commands.py

```diff
@@ -130,3 +130,58 @@ def demux(id_map, fp, out_d, task, maxtask):
     for d in openfps.values():
         for f in d.values():
             f.close()
+
+
+def annotate_filtered_fastq(original_path, stripped_path, output_path):
+    """
+    Annotates a FASTQ file w/missing descriptions using original file.
+    :param original_path: A path to the original, unfiltered file.
+    :param stripped_path: A path to the filtered file.
+    :param output_path: A path for the output file.
+    :return: None
+    """
+    mapping = {}
+
+    # Per FASTQ specification: https://en.wikipedia.org/wiki/FASTQ_format
+    # lines beginning with '@' contain the sequence's identifier and an
+    # optional description or 'metadata'.
+
+    # minimap2 is a tool used to perform host filtering on fastq files.
+    # The preferred output for this program is SAM format:
+    # https://en.wikipedia.org/wiki/SAM_(file_format)
+
+    # SAM format cannot preserve the optional metadata field and hence the
+    # metadata will disappear after using minimap2 to filter a file and
+    # samtools to convert the file from SAM back to FASTQ format.
+
+    # It is straightforward to process even large fastq files line by line
+    # sequence identifier lines are unique and easily identified as they are
+    # the only lines in a FASTQ file beginning with '@'. It is thus straight-
+    # forward to process even large files to build a mapping between
+    # sequence identifiers and optional metadata fields.
+    with open(original_path, 'r') as f:
+        for line in f:
+            if line.startswith('@'):
+                line = line.strip().split()
+                if len(line) != 2:
+                    raise ValueError(f"'{original_path}' does not appear to "
+                                     "contain sequence identifiers with "
+                                     "optional metadata")
+                # where sequence identifier = line[0] and
+                # the optional description = line[1].
+                mapping[line[0]] = line[1]
```

Comment:
For tellseq, we anticipate per-sample files with > 20M records. While this works for small data, this volume of keys will result in the `dict` being quite large. It would be worth testing whether this approach is viable at the anticipated data volume

Reply:
I tested this on the muxed fastq file generated by my iseq run. Based on the 30 seconds it took to process I estimated a run the size of an average NovaSeq run to be about 4 minutes. Just to clarify, is it 20M records across all of the per-sample files or per file? And if so, if the average number of samples was say 300 then we're taking a muxed fastq file of 600M records?

I'll look for an out of core equivalent for a dictionary.

Reply:
Per sample and that would be for R1 only. Our larger samples are > 65M sequences. 

If this is method is intended to be sample agnostic, then the solution needs to be designed to handle a few billion sequences, which is the scale we get from a NovaSeq 6000 lane. 

The issue is primarily space. Remember that hash tables scale poorly. 

I recommend operating per sample, and if the solution does not meet desired resource goals, then to switch to a buffered approach. It is likely that we cannot assume order is preserved, as @antgonza notes. However, we do expect locality.

Alternatively, rather than modifying downstream code, could we instead fix the upstream code so comment lines are retained?

Reply:
The problem is that minimap2 is dropping the optional descriptions when it reads in a fastq file. In a way it makes sense because it outputs in SAM format and the SAM format doesn't appear to support those descriptions. I did look at the code for https://github.com/lh3/minimap2/tree/master for a moment and it seems like the code (specifically in index.c) might be reading in a fixed number of characters to read the sequence identifier instead of reading to newline, as a quick guess.

Hence, modifying the upstream code would involve forking minimap2 or fixing the code and submitting a PR to the owner and hoping they accept it into the main branch. Fixing minimap2 sounds fun and on the plus side they've already got scalability worked out in their memory mapped code. On the downside it's not certain if we can modify it to preserve the description. Updating this solution seems like more of a sure thing, with scaling the top issue. @wasade what do you think we should do?

Reply:
Also, there is Lucas's proposed solution to essentially replace all whitespace with '#' or another character and see if minimap2 accepts it as a name and then we remove it afterward. I'm not certain minimap2 is using a fixed length to extract the string so I'm trying it out now.

Reply:
> Also, there is Lucas's proposed solution to essentially replace all whitespace with '#' or another character and see if minimap2 accepts it as a name and then we remove it afterward. I'm not certain minimap2 is using a fixed length to extract the string so I'm trying it out now.

Sadly this will not work. I took my list of ten sequences and ran them through as is, and four additional times, each time replacing the space in between the sequence id and the description with '_', '-', '#', and '###'. The output for each of the hacked sequences was blank.